presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/PlanChecker.java

```diff
@@ -17,27 +17,41 @@
 import com.facebook.presto.metadata.Metadata;
 import com.facebook.presto.spi.WarningCollector;
 import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.SimplePlanFragment;
 import com.facebook.presto.sql.analyzer.FeaturesConfig;
+import com.facebook.presto.sql.planner.PlanFragment;
+import com.facebook.presto.sql.planner.sanity.plancheckerprovidermanagers.PlanCheckerProviderManager;
 import com.google.common.collect.ImmutableListMultimap;
 import com.google.common.collect.Multimap;
 
 import javax.inject.Inject;
 
+import java.util.List;
+import java.util.stream.Collectors;
+
 /**
  * Perform checks on the plan that may generate warnings or errors.
  */
 public final class PlanChecker
 {
-    private final Multimap<Stage, Checker> checkers;
+    private final PlanCheckerProviderManager providerMgr;
+    private Multimap<Stage, Checker> checkers;
+    private boolean providedInitialized;
```

Comment:
Let's do the following:

1) Let's make `checkers` volatile.  This will be accessed in a multithreaded environment, and the `volatile` keyword will ensure visibility after updates across multiple threads.
2) Let's remove `providedInitialized`, and instead add a method `update` which takes in a `PlanCheckerProvider`, which is called by the `PlanCheckerProviderManager` once the manager is initialized.
3) Now that the flow of communication is reversed, we don't need `ensureProvidedInitialized`.
4) Add a `@GuardedBy("this")` to the class to document the synchronization of this class.

Reply:
I think it would be better to leave this map unmodified, and instead call directly to the planCheckerProviderManager to get any extra checks for each stage.

 The PlanChecker shouldn't have to worry about initialization at all. e.g. if one day we make it possible to dynamically add new plugins, the planChecker shouldn't need to know about that. The PlanChecker should just call out the PlanCheckerProviderManager to know what planCheckers have been registered. 

Follow the example of ConnectorPlanOptimizerManager and how it is used in PlanOptimizers.

But agree with Tim to be wary for any non-final variables (or anything that can be modified, e.g. inserting into a map) you need to make sure that the implementation is thread safe.